File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException{
       BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
       int n = Integer.parseInt(br.readLine());
       int[] dp = new int[n+1];
       int[] a = new int[n+1];
       for(int i =1; i<n+1;i++){
        a[i] = Integer.parseInt(br.readLine());
       }
       if(n > 0){
        dp[1] = a[1];
       }
       if(n>1){
        dp[2] = a[2] + a[1];
       }
       if(n>2){
        dp[3] = Math.max(Math.max(a[3]+a[2],a[3]+a[1]), dp[2]);
       }
       for(int i =4; i<n+1; i ++){
        dp[i] = Math.max(Math.max(a[i]+a[i-1]+dp[i-3], a[i]+dp[i-2]), dp[i-1]);
       }
       System.out.println(dp[n]);
    }
} 

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys
input = sys.stdin.readline
n = int(input())    
a = [0] * (n+1)
dp = [0] * (n+1)
for i in range(1,n+1):
    a[i] = int(input())
if n > 0:
    dp[1] = a[1]
if n > 1: 
    dp[2] = a[2] + a[1]
if n > 2:
    dp[3] = max(a[2]+a[3], a[3]+a[1],dp[2])
    for i in range(4,n+1):
        dp[i] = max(a[i] + dp[i-2], a[i]+a[i-1]+dp[i-3], dp[i-1])
print(dp[n])
